### ISSUE TITLE
ocdav: regression giving 404 after recent config simplification

### ISSUE THREAD

**Issue by PVince81** [owncloud/ocis-reva#270]
It is currently not possible to use Webdav with OCIS / Phoenix any more, getting a 404 error.

I've bisected it and this PR https://github.com/cs3org/reva/pull/826 and more specifically https://github.com/cs3org/reva/pull/826/commits/7d32f1e7e1d647bb91ee8684920d90adb43262e0 is the commit that introduced this issue.

I tried manually created an empty file `/var/tmp/reva/users.json` but it didn't solve the issue, probably because it's unrelated to OCIS user management.

@labkode has anything semantically changed in the Reva configuration ?

**Comment by PVince81** [comment]
More specifically, the request to "remote.php/webdav" is returning a 404.

Logs:
```
==> glauth.log <==
2020-06-15T14:50:24+02:00 DBG Bind request basedn=dc=example,dc=org binddn=cn=konnectd,ou=sysusers,dc=example,dc=org name=glauth service=glauth src={"IP":"127.0.0.1","Port":36058,"Zone":""} verbosity=6
2020-06-15T14:50:24+02:00 DBG Bind success binddn=cn=konnectd,ou=sysusers,dc=example,dc=org name=glauth service=glauth src={"IP":"127.0.0.1","Port":36058,"Zone":""} verbosity=6
2020-06-15T14:50:24+02:00 DBG Search request basedn=,dc=example,dc=org binddn=cn=konnectd,ou=sysusers,dc=example,dc=org filter=(objectClass=posixaccount) name=glauth service=glauth src={"IP":"127.0.0.1","Port":36058,"Zone":""} verbosity=6
2020-06-15T14:50:24+02:00 DBG AP: Search OK filter=(objectClass=posixaccount) name=glauth service=glauth verbosity=6

==> konnectd.log <==
2020-06-15T14:50:24+02:00 DBG  bytes=285 duration=3.226476 method=GET path=/konnect/v1/userinfo proto=HTTP/1.1 request=2755bf27-0d05-4ab6-9407-4e5a7bd2b7e2 service=konnectd status=200

==> proxy.log <==
2020-06-15T14:50:24+02:00 DBG director found path=/remote.php/webdav/ policy=reva prefix=/remote.php/ routeType=prefix service=proxy
2020-06-15T14:50:24+02:00 DBG director found path=/konnect/v1/userinfo policy=reva prefix=/konnect/ routeType=prefix service=proxy

==> reva-auth-bearer.log <==
2020-06-15T14:50:24+02:00 DBG skipping auth method=/cs3.auth.provider.v1beta1.ProviderAPI/Authenticate pkg=rgrpc service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
{"level":"debug","claims":{"email":"einstein@example.org","email_verified":false,"family_name":"Einstein","given_name":"Albert","name":"Einstein","preferred_username":"einstein","sub":"XtSi_miyWSB-pkvGnxPoC5A4flih0UCLgvU7cLwjmjCKX7FYn2HWk6rRCEuy2G5qAy_yMQc_FK9aNFhUMrX2pQ@konnect"},"userInfo":{"sub":"XtSi_miyWSB-pkvGnxPoC5A4flih0UCLgvU7cLwjmjCKX7FYn2HWk6rRCEuy2G5qAy_yMQc_FK9aNFhUMrX2pQ@konnect","profile":"","email":"einstein@example.org","email_verified":false},"time":"2020-06-15T14:50:24+02:00","message":"unmarshalled userinfo"}
2020-06-15T14:50:24+02:00 INF user id:<idp:"https://localhost:9200" opaque_id:"einstein" > username:"einstein" mail:"einstein@example.org" display_name:"Einstein"  authenticated pkg=rgrpc service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 INF unary code=OK end="15/Jun/2020:14:50:24 +0200" from=tcp://127.0.0.1:45846 pkg=rgrpc service=reva start="15/Jun/2020:14:50:24 +0200" time_ns=13715781 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/cs3.auth.provider.v1beta1.ProviderAPI/Authenticate user-agent=grpc-go/1.26.0

==> reva-frontend.log <==
2020-06-15T14:50:24+02:00 WRN core access token not set pkg=rhttp service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 DBG error retrieving credentials error="no basic auth provided" pkg=rhttp service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 DBG credentials obtained from credential strategy: &{Type:bearer ClientID: ClientSecret:eyJhbGciOiJQUzI1NiIsImtpZCI6IiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJwaG9lbml4IiwiZXhwIjoxNTkyMjI1ODE5LCJqdGkiOiJNV1Z4THdUNFQydF9iOTI2dVJ0Y2ZnQUtFNVFMUzNNNyIsImlhdCI6MTU5MjIyNTIxOSwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6OTIwMCIsInN1YiI6Ilh0U2lfbWl5V1NCLXBrdkdueFBvQzVBNGZsaWgwVUNMZ3ZVN2NMd2ptakNLWDdGWW4ySFdrNnJSQ0V1eTJHNXFBeV95TVFjX0ZLOWFORmhVTXJYMnBRQGtvbm5lY3QiLCJrYy5pc0FjY2Vzc1Rva2VuIjp0cnVlLCJrYy5hdXRob3JpemVkU2NvcGVzIjpbImVtYWlsIiwib3BlbmlkIiwicHJvZmlsZSJdLCJrYy5pZGVudGl0eSI6eyJrYy5pLmRuIjoiRWluc3RlaW4iLCJrYy5pLmlkIjoiY249ZWluc3RlaW4sb3U9dXNlcnMsZGM9ZXhhbXBsZSxkYz1vcmciLCJrYy5pLnVuIjoiZWluc3RlaW4ifSwia2MucHJvdmlkZXIiOiJpZGVudGlmaWVyLWxkYXAifQ.LcKrtXwXH2QpIQCYpihNJIz-S_r1QkC3r154D1ojA_lWuLMAzQJeSdFGJyAVYHPxj7ShK2QZx0vaAW4AHe5Is6xJfx7IDzvMjSV0iWhrSyjmHMcevaPXRKtx18alWm3q28i-8ESXh5zZrC4_5b_keB_VoK6aSRxmWVYDZUPsqmSkcCzxiwE7FmV52tRKLQRuXx-6ym9koRlGwAWA6HhmtRKlsisR0OM7HdEeQ2LJfrefanQLlGzmLHLMQlxci0lcdXnzmT8OqvVhOqjGEFbMLadjWhYOoamThHGOjed5E-BXr8MWJvyBN0tNUZ_KW45BMvZKM6hr50HnVfTtXNxKuQ} pkg=rhttp service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 DBG AuthenticateRequest: type:"bearer" client_secret:"eyJhbGciOiJQUzI1NiIsImtpZCI6IiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJwaG9lbml4IiwiZXhwIjoxNTkyMjI1ODE5LCJqdGkiOiJNV1Z4THdUNFQydF9iOTI2dVJ0Y2ZnQUtFNVFMUzNNNyIsImlhdCI6MTU5MjIyNTIxOSwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6OTIwMCIsInN1YiI6Ilh0U2lfbWl5V1NCLXBrdkdueFBvQzVBNGZsaWgwVUNMZ3ZVN2NMd2ptakNLWDdGWW4ySFdrNnJSQ0V1eTJHNXFBeV95TVFjX0ZLOWFORmhVTXJYMnBRQGtvbm5lY3QiLCJrYy5pc0FjY2Vzc1Rva2VuIjp0cnVlLCJrYy5hdXRob3JpemVkU2NvcGVzIjpbImVtYWlsIiwib3BlbmlkIiwicHJvZmlsZSJdLCJrYy5pZGVudGl0eSI6eyJrYy5pLmRuIjoiRWluc3RlaW4iLCJrYy5pLmlkIjoiY249ZWluc3RlaW4sb3U9dXNlcnMsZGM9ZXhhbXBsZSxkYz1vcmciLCJrYy5pLnVuIjoiZWluc3RlaW4ifSwia2MucHJvdmlkZXIiOiJpZGVudGlmaWVyLWxkYXAifQ.LcKrtXwXH2QpIQCYpihNJIz-S_r1QkC3r154D1ojA_lWuLMAzQJeSdFGJyAVYHPxj7ShK2QZx0vaAW4AHe5Is6xJfx7IDzvMjSV0iWhrSyjmHMcevaPXRKtx18alWm3q28i-8ESXh5zZrC4_5b_keB_VoK6aSRxmWVYDZUPsqmSkcCzxiwE7FmV52tRKLQRuXx-6ym9koRlGwAWA6HhmtRKlsisR0OM7HdEeQ2LJfrefanQLlGzmLHLMQlxci0lcdXnzmT8OqvVhOqjGEFbMLadjWhYOoamThHGOjed5E-BXr8MWJvyBN0tNUZ_KW45BMvZKM6hr50HnVfTtXNxKuQ"  against localhost:9142 pkg=rhttp service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 INF core access token generated pkg=rhttp service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 DBG http routing: head=remote.php tail=/webdav svc=not-found pkg=rhttp service=reva
2020-06-15T14:50:24+02:00 WRN http end="15/Jun/2020:14:50:24 +0200" host=127.0.0.1 method=PROPFIND pkg=rhttp proto=HTTP/1.1 service=reva size=0 start="15/Jun/2020:14:50:24 +0200" status=404 time_ns=21903168 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/remote.php/webdav/ url=/remote.php/webdav/

==> reva-gateway.log <==
2020-06-15T14:50:24+02:00 DBG skipping auth method=/cs3.gateway.v1beta1.GatewayAPI/Authenticate pkg=rgrpc service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 DBG skipping auth method=/cs3.auth.registry.v1beta1.RegistryAPI/GetAuthProvider pkg=rgrpc service=reva traceid=bb43310b231fce81f1f42788fc1c3b40
2020-06-15T14:50:24+02:00 INF unary code=OK end="15/Jun/2020:14:50:24 +0200" from=tcp://127.0.0.1:55964 pkg=rgrpc service=reva start="15/Jun/2020:14:50:24 +0200" time_ns=227329 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/cs3.auth.registry.v1beta1.RegistryAPI/GetAuthProvider user-agent=grpc-go/1.26.0
2020-06-15T14:50:24+02:00 INF unary code=OK end="15/Jun/2020:14:50:24 +0200" from=tcp://127.0.0.1:56066 pkg=rgrpc service=reva start="15/Jun/2020:14:50:24 +0200" time_ns=335304 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/cs3.storage.registry.v1beta1.RegistryAPI/GetStorageProvider user-agent=grpc-go/1.26.0
2020-06-15T14:50:24+02:00 INF unary code=OK end="15/Jun/2020:14:50:24 +0200" from=tcp://127.0.0.1:55962 pkg=rgrpc service=reva start="15/Jun/2020:14:50:24 +0200" time_ns=19616995 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/cs3.gateway.v1beta1.GatewayAPI/Authenticate user-agent=grpc-go/1.26.0

==> reva-storage-home.log <==
2020-06-15T14:50:24+02:00 INF unary code=OK end="15/Jun/2020:14:50:24 +0200" from=tcp://127.0.0.1:41690 pkg=rgrpc service=reva start="15/Jun/2020:14:50:24 +0200" time_ns=705970 traceid=bb43310b231fce81f1f42788fc1c3b40 uri=/cs3.storage.provider.v1beta1.ProviderAPI/CreateHome user-agent=grpc-go/1.26.0
```

**Comment by labkode** [comment]
@PVince81 defaults were added for many of the configuration options.

Unless you're using another driver, the file `/var/tmp/reva/users.json` should contain the users that are in the system, can't be empty.

**Comment by labkode** [comment]
Semantically, before we were using `EnableHome` but this now the default and you can disable it with `DisableHome`: https://github.com/cs3org/reva/commit/7d32f1e7e1d647bb91ee8684920d90adb43262e0#diff-814638166b7de5a617bcb3bab07c17b4R53

**Comment by PVince81** [comment]
Thanks for the info, I'll have a look around these

**Comment by PVince81** [comment]
seems the router is not finding that path:
```
2020-06-15T15:00:39+02:00 DBG http routing: head=remote.php tail=/webdav svc=not-found pkg=rhttp service=reva
2020-06-15T15:00:39+02:00 WRN http end="15/Jun/2020:15:00:39 +0200" host=127.0.0.1 method=PROPFIND pkg=rhttp proto=HTTP/1.1 service=reva size=0 start="15/Jun/2020:15:00:39 +0200" status=404 time_ns=28854026 traceid=4962ee72c1e48edc1e2c5f7423da03a9 uri=/remote.php/webdav url=/remote.php/webdav
```

debugging...

**Comment by PVince81** [comment]
current reva master:
```
reva-frontend.log:2020-06-15T15:06:51+02:00 INF http service enabled: datagateway@/data pkg=rhttp service=reva
reva-frontend.log:2020-06-15T15:06:51+02:00 INF http service enabled: ocdav@/webdav pkg=rhttp service=reva
reva-frontend.log:2020-06-15T15:06:51+02:00 INF http service enabled: ocs@/ocs pkg=rhttp service=reva
reva-storage-home-data.log:2020-06-15T15:06:50+02:00 INF http service enabled: dataprovider@/data pkg=rhttp service=reva
reva-storage-oc-data.log:2020-06-15T15:06:50+02:00 INF http service enabled: dataprovider@/data pkg=rhttp service=reva
```

before the simplification commit:
```
reva-frontend.log:2020-06-15T15:11:07+02:00 INF http service enabled: ocdav@/ pkg=rhttp service=reva
reva-frontend.log:2020-06-15T15:11:07+02:00 INF http service enabled: ocs@/ocs pkg=rhttp service=reva
reva-frontend.log:2020-06-15T15:11:07+02:00 INF http service enabled: datagateway@/data pkg=rhttp service=reva
reva-storage-home-data.log:2020-06-15T15:11:07+02:00 INF http service enabled: dataprovider@/data pkg=rhttp service=reva
reva-storage-oc-data.log:2020-06-15T15:11:07+02:00 INF http service enabled: dataprovider@/data pkg=rhttp service=reva
```

so maybe ocis-reva was relying on some default and now we need to set it to an explicit value

**Comment by PVince81** [comment]
it seems it's this line change: https://github.com/cs3org/reva/pull/826/files?diff=unified&w=1#diff-ef5385dd877f5795921538e2e616430aR73

so we need a way to specify/force an empty prefix, I'll try with "/"

**Comment by PVince81** [comment]
as I can see here we already specify an empty prefix: https://github.com/owncloud/ocis-reva/blob/master/pkg/command/frontend.go#L125

I tried replacing with a "/" but now the service path has a double slash:
```
reva-frontend.log:2020-06-15T15:24:51+02:00 INF http service enabled: ocdav@// pkg=rhttp service=reva
```

and routing still fails with 404

So either we need to use another mechanism to detect "no config specified" vs "config value is an empty string", or implement deduplication of slashes for routine.

@labkode thoughts ?

**Comment by PVince81** [comment]
PR here to make empty string the default again: https://github.com/cs3org/reva/pull/837

**Comment by labkode** [comment]
@PVince81 I've merged cs3org/reva#837, that should solve the problem

**Comment by PVince81** [comment]
Thanks a lot!
